### ISSUE TITLE
Cairo runs with normal privileges when configured to run at system startup (Server 2019, UAC)

### ISSUE THREAD

**Issue by lss4** [cairoshell/cairoshell#533]
**TL;DR: When UAC is enabled on Windows Server 2019 (by using a different account than built-in Administrator), Cairo runs with normal privileges when configured to run at system startup, which prevents certain features like Shutdown, Restart from working correctly.**

Original Issue:
---

**Describe the bug**
It seems shutdown/restart doesn't work on Cairo for me. I don't know where I could investigate the cause of it.

This isn't a major issue for the time being as I can always conduct a shutdown/restart through the CTRL-ALT-DEL screen.

**Steps to reproduce**
1. Select Shutdown or Restart from Cairo menu.
2. When asked for confirmation (about unsaved work), click Shutdown/Restart.
3. Nothing happens.

**Expected behavior**
The shutdown/restart in Cairo should work as intended.

**System information:**
 - Windows version: Windows Server 2019 (server equivalent of Windows 10 1809)
 - Cairo version: M3.7174
 - 32-bit or 64-bit: 64-bit
 - In Cairo Settings > Advanced, is Cairo set as shell?: I just enabled it to run at system startup. Not sure if this counts as yes.

**Additional context**
Originally I thought Shutdown Event Tracker might be related. Shutdown Event Tracker is enabled by default on Server editions (Server 2016, 2019, etc.), while disabled on Workstation editions (Windows 10), although this can be configured from Group Policy settings on either edition.

However, it doesn't seem to be the root cause as I still cannot shutdown/restart from Cairo even after having it disabled through Group Policy settings (that it doesn't ask me for a reason before actual shutdown/restart anymore).

**Screenshots**
N/A

**Comment by dremin** [comment]
Interesting, it works for me on Hyper-V Server 2019. Do you have any security software which might be blocking this from non-system processes?

**Comment by lss4** [comment]
> Interesting, it works for me on Hyper-V Server 2019. Do you have any security software which might be blocking this from non-system processes?

I don't use any security software other than what Microsoft already provides.

Also, I don't use the Administrator account directly. I created a different Administrator account and disabled the built-in one (this way UAC would be active). It's possible that Cairo is not running with full system rights since UAC is enabled here...

I can always conduct shutdown/reboot from CTRL-ALT-DEL screen.

**Comment by dremin** [comment]
@lss4 There is a log but I don't think that system command produces any log output. You can access the log via Cairo Settings > Advanced.

I haven't tested with a separate admin account on Server, so I will check out that scenario.

**Comment by lss4** [comment]
> There is a log but I don't think that system command produces any log output. You can access the log via Cairo Settings > Advanced.
> 
> I haven't tested with a separate admin account on Server, so I will check out that scenario.

You're right about this... I tried setting log level to Debug and nothing related to Shutdown/restart showed up in the log, only some records of redrawing.

Maybe I'll take a look at how Cairo conducts shutdown/restart and try figuring it out...

**Comment by lss4** [comment]
UPDATE: Okay I've found out the reason. It seems Cairo is indeed being run with normal user privileges in my case, that `ExitWindowsEx` (used for shutdown and restart) cannot function.

If I exit Cairo and relaunch it via "Run as Administrator", Shutdown/Restart works as intended.

So it appears when I configured Cairo to run at system startup, it was not being configured to run with Adminstrator privileges so this issue surfaced.

**Comment by dremin** [comment]
I was able to reproduce this using a non-administrator account, but that is because the non-administrator accounts by default do not have permission to shut down the system. Does any other application, such as the `shutdown` command work for you?

**Comment by lss4** [comment]
> I was able to reproduce this using a non-administrator account, but that is because the non-administrator accounts by default do not have permission to shut down the system. Does any other application, such as the `shutdown` command work for you?

That command needs to be explicitly run as Administrator in order to work (which is intended). The same applies to the `ExitWindowsEx` call.

In my case, my account _is_ an Administrator. It's just that UAC is active so it runs mostly with standard privileges, and I would be prompted if something needs elevated privileges.

However, it seems to cause Cairo to run with only standard privileges by default, and it doesn't even at least ask me for elevation.

**Comment by dremin** [comment]
In `secpol.msc` > Local Policies > User Rights Assignment > Shut down the system, if you add your user, then log out and back in, do these features work in Cairo without elevating Cairo?

**Comment by lss4** [comment]
> In `secpol.msc` > Local Policies > User Rights Assignment > Shut down the system, if you add your user, then log out and back in, do these features work in Cairo without elevating Cairo?

Wow, you're right. Adding my user there worked.

Since on default Windows shell everything was working as intended I never noticed this would be the cause of my issue.

Many thanks! I've learned something useful today.
